### PR TITLE
[In progress] Add warning padding attention mask

### LINKED ISSUE #16136: Add warning message if model uses `input_ids` that include padding tokens, but no `attention_mask` is provided.
## **First good issue** 

A current error is that a user forwards a batched tensor of `input_ids` that include a padding token, e.g. ```input_ids = torch.tensor([["hello", "this", "is", "a", "long", "string"], ["hello", "<pad>", "<pad>", "<pad>", "<pad>"]]```

In this case, the `attention_mask` should be provided as well. Otherwise the output hidden_states will be incorrectly computed. This is quite a common silent error IMO. 

With @LysandreJik @sgugger, we have decided to not **automatically** create the `attention_mask` that masks out the padding tokens in this case because of the reasons explains here: https://github.com/huggingface/transformers/issues/15479#issuecomment-1066639938 . However as pointed out in https://github.com/huggingface/transformers/issues/15479, we should IMO at least displa a warning since this error happens a lot IMO. 

As a first good issue, one could add such a warning to the BertModel in a first case which would go something like:

```py
if attention_mask is not None and (input_ids == pad_token_id).any():
    logger.warn("display nice warning here....")
```

What do you think @sgugger @LysandreJik ?

### PR DESCRIPTION
**Pull request by anruijian**
# What does this PR do?

Fixes #16136 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by sgugger** [comment]
cc @gante

**Review by gante — COMMENTED** [review]
I share the concerns of @sgugger stated in [this comment](https://github.com/huggingface/transformers/issues/16136#issuecomment-1066839556) -- If the user is not passing the attention mask, but doing things right, all these checks will be done at each forward pass, probably causing unnecessary slowdowns.

Why not a simple `logger.warning_once()` if  `attention_mask is None`?  It is guaranteed to run only once at train time, is much shorter in terms of code, and accounts for special tokens that may exist in future models.

**Comment by ydshieh** [comment]
> Why not a simple logger.warning_once()

This is recently introduced :-)

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1043,6 +1043,72 @@ def get_input_embeddings(self) -> nn.Module:
...
+        if attention_mask is not None:
+            # if the attention_mask is defined it's all good
+            return
+
+        if not hasattr(self, "warnings_issued"):
+            self.warnings_issued = {}
+
+        if self.warnings_issued.get("pad_token_in_input_ids", False):
+            # if warning has already been thrown don't throw again
+            return
```
Comment (ydshieh) [inline src/transformers/modeling_utils.py:1056]: This makes the warning appear once, cc @gante . But it's nice if we can re-use the recently added `logger.warning_once`. I will leave @gante to make the final call.

**Comment by gante** [comment]
@anruijian I see, you are absolutely right, it would only do the check in the first forward pass :D 

Still, if the first batch is correct by luck and subsequent ones aren't, no warning would be issued (undesirable). So we would need even more checks if we go down this route. For all these reasons, I'd strongly advocate for a simple warning when the attention mask is not passed, sticking to the KISS principle 🤗 

As a side note, we have related problems at other points in the code base. Getting into the habit of passing the `attention_mask` would really make everyone happier!

**Comment by gante** [comment]
@anruijian It checks `input_ids` until there is a batch where a `pad_token_id` exists. If a user is working on a problem where  they have no `pad_token_id` on their data, there is a check made every iteration. I'd strongly advocate for a simple warning when the attention mask is not passed, sticking to the KISS principle 🤗 

As a side note, we have related problems at other points in the code base. Getting into the habit of passing the `attention_mask` would really make everyone happier!

**Comment by anruijian** [comment]
@gante Just to confirm before updating the PR, we are going to remove `warn_if_pad_token_in_input_ids_no_attention_mask` method and use `logger.warning_once` in `forward()`:
```python
def forward(...):
    ...
    if not attention_mask:
        logger.warning_once(
            "\nWe strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the"
            " attention weights. "
        )
    ...
```

**Comment by gante** [comment]
@anruijian correct :) I would add a short example in the warning, such as `(e.g. to correctly mask the pad tokens)`, but I'll leave that up to you!

**Comment by anruijian** [comment]
@gante 
```python
def forward(...):
    ...
    if not attention_mask:
        logger.warning_once(
            "\nWe strongly recommend passing an `attention_mask` to avoid possibly incorrectly computing the"
            " attention weights. Example to correctly mask the pad tokens: model(input_ids, attention_mask=attention_mask)."
            " See https://huggingface.co/docs/transformers/v4.23.1/en/troubleshooting#incorrect-output-when-padding-tokens-arent-masked for more details."
        )
    ...
```
Does this example look good to you? I also link the official doc on the issue. Not sure if it's too long. Let me know what you think about this. Thanks!
